src/View/Parsers/HtmlDiff.php

```diff
@@ -0,0 +1,288 @@
+<?php
+
+namespace SilverStripe\View\Parsers;
+
+use Masterminds\HTML5\Elements;
+use SebastianBergmann\Diff\Differ;
+use SilverStripe\Core\Convert;
+use SilverStripe\Core\Injector\Injector;
+
+/**
+ * Class representing a 'diff' between two sequences of HTML strings.
+ */
+class HtmlDiff
+{
+    private const OLD_VAL = 'old';
+    private const NEW_VAL = 'new';
+
+    private static ?Differ $differ = null;
+
+    public static ?string $html_cleaner_class = null;
+
+    /**
+     * Get a diff between two sets of HTML content. The result is an HTML fragment which can be added directly
+     * into the DOM. <ins> elements are used to indicate new content, and <del> elements are used to indicate
+     * removed content.
+     *
+     * @param bool $escape If true, the HTML in $from and $to will be escaped after the diff operation is performed.
+     */
+    public static function compareHtml(string|array $from, string|array $to, bool $escape = false): string
+    {
+        // Get HTML chunks even if we're going to escape it later
+        // The diff algorithm sees "<span>some" as a single piece rather than "<span>" and "some" being separate
+        $from = self::explodeToHtmlChunks($from);
+        $to = self::explodeToHtmlChunks($to);
+
+        // Diff the chunks
+        $differ = self::getDiffer();
+        $diff = $differ->diffToArray($from, $to);
+
+        // If we aren't escaping the HTML, convert the first diff into clean HTML blocks and then run a new diff
+        // on the blocks to get an end result that doesn't have broken HTML
+        if (!$escape) {
+            $diifAsBlocks = self::convertDiffToHtmlBlocks($diff);
+            $diff = $differ->diffToArray($diifAsBlocks[self::OLD_VAL], $diifAsBlocks[self::NEW_VAL]);
+        }
+
+        $diff = self::createFinalDiffBlocks($diff, $escape);
+
+        // Take the diff and slap the appropriate <ins> and <del> tags in place
+        $content = '';
+        foreach ($diff as $edit) {
+            list($value, $type) = $edit;
+            if (is_array($value)) {
+                $value = implode(' ', $value);
+            }
+            if ($escape) {
+                $value = Convert::raw2xml($value);
+            }
+
+            switch ($type) {
+                case Differ::OLD:
+                    $content .= ' ' . $value . ' ';
+                    break;
+
+                case Differ::ADDED:
+                    $content .= ' <ins>' . $value . '</ins> ';
+                    break;
+
+                case Differ::REMOVED:
+                    $content .= ' <del>' . $value . '</del> ';
+                    break;
```

Comment:
Oops sorry, I forgot to reply to this. The type will always be one of the `Differ` constants. See https://github.com/sebastianbergmann/diff/blob/70dd1b20bc198da394ad542e988381b44e64e39f/src/Differ.php#L73-L111

The only exception would be if there are unmatched line endings I guess which I hadn't noticed before - I'll add a test for that and add a case if the test indicates one is needed (it's still one of the constants, just not one we're handling at the moment)

Reply:
I've ended up adding both an exception for anything we don't expect (because there's no clear way to handle those, but they shouldn't happen anyway) and a new test which makes it clear that the `Differ::DIFF_LINE_END_WARNING` constant shouldn't ever be a problem for our use case.